### PR TITLE
[Duplicate of #7617] - Next.js basics with payroll system

### PR DESCRIPTION
**Pull request by unholydisaster**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [ ] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [ ] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [ ] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [ ] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [ ] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [ ] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [ ] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [ ] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [ ] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [ ] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [ ] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [ ] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [ ] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [ ] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by unholydisaster** [comment]
Am I still on the right track, I drafted the PR according to your instruction.

**Comment by srishilesh** [comment]
I'll do a tech review ASAP. Will let you know if there are any more suggestions.

**Comment by unholydisaster** [comment]
Yoh, what's up with the review

**Comment by srishilesh** [comment]
Hey,

Will review it asap.

**Comment by unholydisaster** [comment]
Hey, is something wrong with the article ?

**Comment by srishilesh** [comment]
Hey,

I'll review it soon. I have been busy for the past few weeks.

**Comment by srishilesh** [comment]
The review will be done by this weekend.

**Comment by unholydisaster** [comment]
Thanks 👍

**Comment by srishilesh** [comment]
Review is in-progress

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
+In this tutorial, we learn basics of Next.js and how we can use them effectively to build a payroll system.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:1]: Can you add an intro about NextJS? Where we use it? How it is better than React? What is it?

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+- [Reference](#reference)
+
+### Prerequisites
+To follow along, the reader should have the following:
+- Basic knowledge of Javascript.
+- Basic knowledge of React.js and Next.js.
+- Fully set development environment.
+
+### Next.js environment setup
+In this tutorial, we will use Node.js will be needed for the javascript runtime needed.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:21]: Why are we using NodeJS? Please mention the reason

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+You can download Node.js [here](https://nodejs.org/en/download/) if you don't have it installed.
+Type this in a fresh command-line interface to install Next.js.
+
+```bash
+npx create-next-app@latest
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:27]: Shouldn't you specify the name of the app here?

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+`Net pay=Gross pay - Income tax - Health insurance`
+
+
+### Creating a JSON File and Functions in Next.js
+
+This is the project workspace created earlier.
+
+![VSWorkspace](/engineering-education/building-a-payroll-system-with-nextjs/folder-structure.png)
+
+We need timesheet data but in our case, we are going to create it locally and store it in our file system.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:60]: Please use ` `` ` for keywords, file names, variables and so on

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+  Houseallowance:50,
+  Conveyanceallowance:40,
+  Childreneducationallowance:50,
+  FuelReimbursements:50,
+  DriverReimbursements:100,
+  HealthInsurance:10,
+  Taxincome:10
+}]
+```
+In order for this to work as planned, we will have to create an API route for data fetching. Navigate to the Pages folder and open the API folder that contains a hello.js file, edit the file by erasing all its contents and copy this code.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:121]: What methods are we building? What do they do?

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+  Employed.push({ ...body, id:Employed.length +1});
+  return res.status(200).json(Employed);
+}
+}
+
+```
+This creates an API route with methods that are able to send and receive "GET" and "POST" requests for our body that we created earlier by importing it and sending it as a response to our required file.
+
+After dealing with the API route, we are going to create a function that computes all this data by automatically linking it to the arithmetic calculations.
+We are going to use a .map() function to iterate our data and create a list from this data.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:144]: You can state them above the code snippets

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+  const TotalReimbursement=Data.FuelReimbursements +Data.DriverReimbursements
+
+  const Totaldeduction=Data.HealthInsurance +Data.Taxincome
+
+  const Grosspay=Basicpay + Totalallowance + TotalReimbursement
+  const Netpay=Grosspay-Totaldeduction
+                 return(
+                <>
+                <div key={Data.Fullname}>
+                <Link href={`/${Data._Fullname}`} passHref>
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:210]: Code needs formatting

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+                </>
+                 )
+              })}
+      </>       
+    )
+}
+export async function getServerSideProps(){
+    
+    const data=await fetch(`http://localhost:3000/api/hello/`);
+    const employed=await data.json()
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:225]: Explain

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    }
+    return{
+        props:{
+            employees:employed
+        }
+    }
+}
+
+export default Payroll
+```
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:240]: Explain the above code as bulleted points

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @unholydisaster ,

Apologies for the delay in review.

On going through your tutorial, I find that we might need to work on some major improvements. I've added a few suggestions. I would also advise you to take a look at other published articles to see how the authors have structured the flow of the article.

Don't just state the code snippets, instead explain them.

Make sure to do all the changes and ping me when you are done.

Thanks.

**Comment by unholydisaster** [comment]
Hey, finished all changes requested 👍

**Comment by WanjaMIKE** [comment]
Plagiarism check is done

**Comment by unholydisaster** [comment]
Hey, all suggestions done ✅

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+In this tutorial we will use Next.js, Node.js will also be needed for creation of a server which provides a runtime environment which is essential to program debugging and development.
+You can download Node.js [here](https://nodejs.org/en/download/) if you don't have it installed.
+Type this in a fresh command-line interface to install Next.js.
+
+```bash
+npx create-next-app@latest
+```
+![Installation](/engineering-education1/next.js-basics-with-payroll-system/image-three.png)
+
+The installation will begin and you will be asked to name the application, let's call it payroll.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:32]:
```suggestion
The installation will begin and you will be asked to name the application, let's call it `payroll`.
```

File: content/articles/building-a-payroll-system-with-nextjs/index.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+```bash
+cd payroll
+npm run dev
+```
+The server will start at port 3000, click on the provided link to view it on the browser.
+ 
+
+
+### Introduction to Payroll System
+A payroll system is a process of calculating employees' pay for work performed during a specific period. In search a system, a timekeeping system is required to approve the hours worked, Overtime and wages.
```
Comment (srishilesh) [inline content/articles/building-a-payroll-system-with-nextjs/index.md:44]: Is a process?

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @unholydisaster ,

On going through the tutorial, I found that it is a little short on content. We have other tutorials that introduces the reader to NextJS. I would suggest you to improve the existing article by adding more content that are valuable or build a full working application (do not break down this implementation into the next tutorial).

Also, perform a grammar check. Feel free to go through my articles to see how I have formatted them.

Thanks.

**Comment by hectorkambow** [comment]
Our team has been working diligently to clear the backlog and we should be getting this pull request ready to be published next week 🚀

**Comment by srishilesh** [comment]
Hey @unholydisaster ,

I don't see any latest commits. Did you forget to push them ?

**Comment by unholydisaster** [comment]
I will push them today.

**Comment by unholydisaster** [comment]
Hey, suggestions done ✅

**Comment by hectorkambow** [comment]
@unholydisaster 

Hello Eli - and thank you for submitting your 1st article to be published. 

Please keep in mind that we ask students only to submit polished and ready-to-be-published articles. 

Articles should be proof-read and the code should be already reviewed to increase the review process. 

EngEd Peer Reviewers are here to help make sure there are not any GLARING errors and to close the pull request if needed. 

These articles are public-facing and students should keep in mind that it will reflect how the larger DEV community will view your work/projects/skills. - and can impact your software career journey 🚀 .

**Comment by unholydisaster** [comment]
Thanks alot 👍

**Comment by hectorkambow** [comment]
@unholydisaster 

Eli please be sure to review and go over any and all edits suggested by your peer reviewer or made by myself - to improve future articles/blog posts. 

You'll be able to see your live published article on the site (in about 30 mins or so).

To give your article a boost in page engagement - be sure to share it across your social site (LinkedIn/Twitter) and tag the EngEd program so we can also share it and give it some traction/exposure as well!

I believe this is your 1st EngEd article that was published - you may have forgot to add/create an author page to go alongside your articles (as instructed in the EngEd docs).

NOTE: Also, take into consideration that you can add any other past projects of yours: videos, talks, tutorials, articles (on ANY other site) - you can add them to your EngEd profile.

If you have not done so already - consider adding a skill section to your author page. This will go a long way for potential future employers.

Think of the EngEd profile as your one-stop shop for everything you do in DEV - making it easier for you to share all your accomplishments when you go out for an interview or employment opp.

It can be from any blog site (or site in general).

Let me know if you have any questions moving forward.

**_Also important NOTE: We wanted to mention that we're going to take a focus on content centered around Kubernetes (K8s) for the month of August._**

The best five (5) K8s related articles that come in this month will be eligible to receive a complimentary virtual pass to [KubeCon USA](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/) ($75.00 value) in Oct.

We're looking for content that enriches developers' understanding of K8s, best practices, and its many use cases.
Please be sure to go over the already [published content the EngEd community has created on Kubernetes](https://www.section.io/engineering-education/search/?q=kubernetes) to see where we can build upon or further explore. 🚀

**Comment by unholydisaster** [comment]
Thanks alot Hector

**Comment by unholydisaster** [comment]
It's my rightful duty to follow all suggestions.

**Comment by unholydisaster** [comment]
Is there a way, that I can add an author's page to the article.

**Comment by hectorkambow** [comment]
That was my mistake - I did not see the author file. 

I've went ahead and added it for you.  & also created a skill section for you (feel free to add any skills you see fit)

**Comment by unholydisaster** [comment]
Thanks 👍, God bless you.

**Pull request merged**